### ISSUE TITLE
For any reason, 1 specific PNG fails in 4 channel mode, but succeeds with 3 channels

### ISSUE THREAD

**Issue by manticore-projects** [veluca93/fpnge#31]
Greetings!

While everything is working great in general, I found one image, that segfaults in 4 channel mode but succeeds in 3 channel mode:

malloc(): invalid size (unsorted)
malloc(): corrupted top size

Any idea why and how to fix it? I could not spot anything exceptional on this image:
Image: ![failure16044445656400295224](https://github.com/user-attachments/assets/9d6dc724-a421-4a1d-9a48-5ed21faa0483)

Interestingly this also fails with the SSE only version.

**Comment by manticore-projects** [comment]
I have opened/saved this file with GIMP. The newly written file works without any problem. But no other tool complains about this PNG being invalid.

**Comment by veluca93** [comment]
I do not seem to have any issues - can you tell me more about how you're reproducing the problem?

**Comment by manticore-projects** [comment]
Thank you for returning back so quick! The problem happens when wrapping into java. I am calling the following code:

```c
    EXPORT ByteArray* fpng_encode_image_to_memory(const void* pImage, uint32_t w, uint32_t h, uint32_t num_chans,  uint32_t flags = 0) {
        // Vector frees itself
        std::vector<uint8_t> out_buf;

        printf("Program started...\n"); // Check if the program starts correctly
        fflush(stdout);                 // Ensure output is flushed

        if (num_chans==(uint32_t) 4) {
            swapChannelsABGRtoRGBA( (unsigned char*) pImage, w * h);
        } else {
            swapChannelsBGRtoRGB( (unsigned char*) pImage, w * h);
        }

        bool result = fpng::fpng_encode_image_to_memory( pImage, w, h, num_chans, out_buf, flags);

        printf("Size of out_buf: %zu\n", out_buf.size());
        fflush(stdout);

        ByteArray* data = (ByteArray*) malloc( sizeof(ByteArray) );
        data->size = out_buf.size();
        data->data = (uint8_t*) malloc( out_buf.size() * sizeof(uint8_t) );

        // copy the uint8_t* array
        std::memcpy(data->data, out_buf.data(), out_buf.size());

        return data;
    }
```

This works in general except with this File and 4 channels. Even 3 channels work with this file.
I have no idea why MALLOC would fail here.

**Comment by manticore-projects** [comment]
The full java project is here: https://github.com/manticore-projects/fpng-java
The relevant source is here: https://github.com/manticore-projects/fpng-java/blob/cd7e9e727d5e102bc0cd48cd5b6d39e52fa456e5/fpnge/src/main/cpp/fpnge.cc#L1617

**Comment by manticore-projects** [comment]
I think, it's this one: 
```c
data->data = (unsigned char*) malloc(data->size);
```
But why and how?

**Comment by veluca93** [comment]
Are you able to run your project under address sanitizer?

**Comment by manticore-projects** [comment]
> Are you able to run your project under address sanitizer?

Sorry, I am just a Java programmer and have no idea how to achieve this.
And first of all, I would love to understand why/how this MALLOC can fail. I though it would just allocate memory of a certain size?

**Comment by veluca93** [comment]
The problem is not malloc failing, but some memory corruption happening somewhere else.

I am not an expert on JNI bindings. When building regular C/C++ code, passing `-fsanitize-address` enables address sanitizer. I also see that for Android NDK one can enable address sanitizer as explained [here](https://developer.android.com/ndk/guides/hwasan)

**Comment by manticore-projects** [comment]
> -fsanitize-address

Sure, I can do do this, it goes to the GCC options, right?

**Comment by manticore-projects** [comment]
This are my GCC options now:

```
['-fPIC', '-O3',  '-march=x86-64', '-mtune=generic', '-funroll-loops', '-std=c++17', '-Wall', '-msse4.2', '-mpclmul', '-msse2', '-mavx2', '-fno-strict-aliasing', '-fsanitize=address']
```

Did not change anything so.
The problem is, that this segfaults the JVM and so I can't see anything, except this "malloc(): corrupted top size" error.

**Comment by veluca93** [comment]
Are you able to reproduce this issue outside of a JVM environment?

**Comment by manticore-projects** [comment]
No. I tested the plain original libraries and it worked.
I am very sure that I do something wrong. I am just lost where exactly to look.

**Comment by StefanVK** [comment]
@manticore-projects I believe there may a problem with _swapChannelsABGRtoRGBA_
```
extern "C" EXPORT void swapChannelsABGRtoRGBA(unsigned char* pImage, int numPixels) {
    const __m128i shuffleMask = _mm_set_epi8( 12, 13, 14, 15, 8, 9, 10, 11, 4, 5, 6, 7, 0, 1, 2, 3);

    for (int i = 0; i < numPixels; i += 4) {
        __m128i abgr = _mm_loadu_si128((__m128i*)(pImage + i * 4));
        __m128i rgba = _mm_shuffle_epi8(abgr, shuffleMask);
        _mm_storeu_si128((__m128i*)(pImage + i * 4), rgba);
    }
}
```

It handles 4 Pixels at a time. But your image is 21x17 = 357 Pixels. So it will shuffle 3 Pixels or 96 bytes past your image buffer, no?

When using SIMD intrinsics or instructions you should usually check that you have enough data left to handle a full set and fall back to an algorithm that handles single elements at a time for the remainder.

Something like e.g.
```
extern "C" EXPORT void swapChannelsABGRtoRGBA(unsigned char* pImage, int numPixels) {
    const __m128i shuffleMask = _mm_set_epi8(12, 13, 14, 15, 8, 9, 10, 11, 4, 5, 6, 7, 0, 1, 2, 3);

    int i;
    for (i = 0; i + 3 < numPixels; i += 4) {
        __m128i abgr = _mm_loadu_si128((__m128i*)(pImage + i * 4));
        __m128i rgba = _mm_shuffle_epi8(abgr, shuffleMask);
        _mm_storeu_si128((__m128i*)(pImage + i * 4), rgba);
    }

    // Handle remaining pixels
    for (; i < numPixels; ++i) {
        const unsigned char a = pImage[i * 4];
        const unsigned char b = pImage[i * 4 + 1];
        const unsigned char g = pImage[i * 4 + 2];
        const unsigned char r = pImage[i * 4 + 3];
        
        pImage[i * 4] = r;
        pImage[i * 4 + 1] = g;
        pImage[i * 4 + 2] = b;
        pImage[i * 4 + 3] = a;
    }
}
```

I don't know if this is the cause of your crash. But I think it's worth taking a look it.

**Comment by manticore-projects** [comment]
@StefanVK: you are my hero and I tip my hat! Thank you!

There were a few issues with my code (and I was too tired yesterday, to think it through methodically). Be set for a good laugh:

1) it really confused me that the error was MALLOC related (a java programmer gets always an exception with a line number and so he is easily lost with C error behaviour)

2) that's why I did not see into the Byte Swap methods (and you are totally right of course)

3) now the most ridiculous part: Since I need to carry the C code wrapped in the Java library and unrwap it, I implemented a caching mechanism and forgot about it. So whenever I ran my tests and tried to debug, this cache was touched and never the actual work. Funny me.

Apologies for wasting your time, thank you much for your help and cheers!
Andreas
